hswidget/cof_widget.go

```diff
@@ -0,0 +1,668 @@
+package hswidget
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/ianling/giu"
+	"github.com/ianling/imgui-go"
+
+	"github.com/OpenDiablo2/OpenDiablo2/d2common/d2enum"
+	"github.com/OpenDiablo2/OpenDiablo2/d2common/d2fileformats/d2cof"
+
+	"github.com/OpenDiablo2/HellSpawner/hscommon/hsenum"
+)
+
+const (
+	indicatorSize                        = 64
+	buttonWidthHeight                    = 15
+	actionButtonW, actionButtonH         = 200, 30
+	saveCancelButtonW, saveCancelButtonH = 80, 30
+	bigListW                             = 200
+	trueFalseListW                       = 60
+	speedInputW                          = 40
+)
+
+type cofTextures struct {
+	up    *giu.Texture
+	down  *giu.Texture
+	left  *giu.Texture
+	right *giu.Texture
+}
+
+// COFWidget represents cof viewer's widget
+type COFWidget struct {
+	id       string
+	cof      *d2cof.COF
+	textures cofTextures
+}
+
+// COFViewer creates a cof viewer widget
+func COFViewer(
+	up, down, right, left *giu.Texture,
+	id string, cof *d2cof.COF,
+) *COFWidget {
+	result := &COFWidget{
+		id:  id,
+		cof: cof,
+	}
+
+	result.textures.up = up
+	result.textures.down = down
+	result.textures.left = left
+	result.textures.right = right
+
+	return result
+}
+
+func (p *COFWidget) getState() *COFState {
+	stateID := fmt.Sprintf("COFWidget_%s", p.id)
+	currentState := giu.Context.GetState(stateID)
+
+	if currentState == nil {
+		currentState = p.setupDefaultState(stateID)
+	}
+
+	return currentState.(*COFState)
+}
+
+// Build builds a cof viewer
+func (p *COFWidget) Build() {
+	state := p.getState()
+
+	// builds appropriate menu (depends on state)
+	switch state.mode {
+	case cofEditorModeViewer:
+		p.makeViewerLayout().Build()
+	case cofEditorModeAddLayer:
+		p.makeAddLayerLayout().Build()
+	case cofEditorModeConfirm:
+		giu.Layout{
+			giu.Label("Please confirm your decision"),
```

Comment:
we can consider hidding main window, when PopUpConfirmDialog is open

Reply:
yeah, that sounds like a good idea.

Reply:
opened an issue for this #195